### DISCUSSION TITLE
Metals does not include AWS profiles

### DISCUSSION THREAD

**Discussion by bbstilson** [scalameta/metals#2619] [Q&A] [answered]
I am not sure what the right solution here would be, so I will explain my situation and perhaps someone can provide a workaround. I wrote [a plugin](https://github.com/bbstilson/sbt-codeartifact) that uses the AWS SDK to add a CodeArtifact repository as a resolver. In order to use the plugin, the user must have AWS Credentials available in some way ([there are many ways](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/auth/credentials/DefaultCredentialsProvider.html)). Unfortunately, metals doesn't seem to have access to them, or isn't providing them, which means any build that uses this plugin will not work.

Happy to explain further. Thanks.

**Comment by tgodzik** [comment]
Metals doesn't download any artifacts by itself (except Metals artifacts themselves) and everything should work normally via sbt. What is the exact issue you are encountering here? It would be useful to see at which point you are encountering an exception.

**Reply by bbstilson** [reply]
For example, one of the ways in which the aws sdk will load credentials and configuration is via a credentials and configuration files located at `~/.aws/{credentials,configuration}`. When I start sbt manually from my terminal (either in vscode or iterm), those credentials are discoverable. Metals, however, fails to load my sbt project because the plugin throws an exception at startup because the aws credentials are not available.

**Reply by bbstilson** [reply]
The full stack trace is not helpful, but this is the main error:

```
awssdk.regions.providers.DefaultAwsRegionProviderChain@36ce901d: [software.amazon.awssdk.regions.providers.SystemSettingsRegionProvider@28d03787: Unable to load region from system settings. Region must be specified either via environment variable (AWS_REGION) or  system property (aws.region)., software.amazon.awssdk.regions.providers.AwsProfileRegionProvider@21e2bcd: No region provided in profile: default, software.amazon.awssdk.regions.providers.InstanceProfileRegionProvider@374ec881: Unable to contact EC2 metadata service.]
```

The key is `No region provided in profile: default`. I don't have a default profile. Instead I set one via envvars. So it seems like metals is starting sbt in some context which doesn't have those envvars.

**Reply by tgodzik** [reply]
I think it doesn't have the env vars unless you start VS Code in the terminal with those variables specified. You should be able to specify .sbtopts (or .jvmopts) file that can have proper java properties specified and that should always be picked up.

**Reply by bbstilson** [reply]
I was able to add a custom start up script that sources my envvars. Thanks for your help!

**Comment by dos65** [comment, marked as answer]
@bbstilson if your setup relies on specific environment variables you may find some hints in this thread - https://github.com/scalameta/metals/discussions/2418

**Reply by bbstilson** [reply]
Hey, thanks this helped!

For those who don't want to do more digging, I added a shell script here: `/Users/home/.vscode/metals/startup.sh` with the following contents:

```bash
#!/usr/local/bin/bash

source ~/.zprofile
sbt "$@"
```

Then in the `Sbt script` vscode metals setting, I added that full path: `/Users/home/.vscode/metals/startup.sh`.
